oak_idl_gen_services/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+/// - a struct named `NameServer`, which implements the `oak_idl::Transport` trait, dispatching each
+///   request to the appropriate method on the underlying service implementation.
+/// - a trait named `Name`, with a method for each method defined in the macro, and an additional
+///   default method named `serve` which returns an instance of `NameServer`; the developer of a
+///   service would usually define a concrete struct and manually implement this trait for it.
+///
+/// For an input flatbuffer file with name `foo.fbs`, the generated Rust file will be located at
+/// `${OUT_DIR}/foo_services.rs`.
+pub fn compile_services(filename: &str) {
+    // Run flatc to generate the reflected schema in flatbuffer binary format.
+    let output = std::process::Command::new("flatc")
+        .args([
+            "--schema",
+            "--binary",
+            "-o",
+            &std::env::var("OUT_DIR").unwrap(),
+            filename,
+        ])
+        .output()
+        .unwrap();
+    if !output.status.success() {
+        eprintln!("flatc exit code: {}", output.status);
+        eprintln!(
+            "flatc stdout: {}",
+            std::str::from_utf8(&output.stdout).unwrap()
+        );
+        eprintln!(
+            "flatc stderr: {}",
+            std::str::from_utf8(&output.stderr).unwrap()
+        );
+        exit(1);
+    }
+    // We then read back the generated schema and parse it with the reflection flatbuffer schema.
+    // See https://google.github.io/flatbuffers/intermediate_representation.html.
+    let file_prefix = std::path::Path::new(filename)
+        .file_prefix()
+        .unwrap()
+        .to_str()
+        .unwrap();
+    let schema_file = format!("{}/{}.bfbs", std::env::var("OUT_DIR").unwrap(), file_prefix);
+    let schema_bytes = std::fs::read(&schema_file).unwrap();
+    let generated_schema = generate_from_bytes(&schema_bytes).unwrap();
+    let out_file = format!(
+        "{}/{}_services.rs",
+        std::env::var("OUT_DIR").unwrap(),
+        file_prefix
+    );
+    std::fs::write(&out_file, generated_schema).unwrap();
+}
+
+/// Generate the Rust file from the input flatbuffer file.
+///
+/// In general, each flatbuffer file may have zero or more `rpc_service` instances defined.
+fn generate_from_bytes(schema_bytes: &[u8]) -> anyhow::Result<String> {
+    let schema = reflection_generated::reflection::root_as_schema(schema_bytes)
+        .context("could not parse schema")?;
+    Ok(schema
+        .services()
+        .context("could not find any service definition")?
+        .iter()
+        .map(|s| generate_service(&s))
+        .collect::<Result<Vec<_>, _>>()?
+        .into_iter()
+        .intersperse("\n".to_string())
+        .collect())
+}
+
+/// Generate the Rust objects from the input [`Service`] instance, corresponding to an `rpc_service`
+/// entry.
+fn generate_service(service: &Service) -> anyhow::Result<String> {
+    let mut lines = Vec::new();
+    lines.extend(vec![
```

Comment:
I'm surprised we can't use a rust macro, might be helpful to explain the choice in the comment briefly. :)

Reply:
I am still a bit worried about maintaining/changing the code printed out by `format`. 

I can't come up with a better suggestion, but I think it is worth to investigate soon. With macros, as Juiette said or also, e.g., there seems to be [codegen](https://docs.rs/codegen/0.1.3/codegen/).

I am not sure what you considered already, maybe it is worth adding a bried comment on your decision.

Is the code generated in the CI? Then we will note if there is some problem with the code, e.g., a new linting rule fires, yes? 

In general we could not do identation here, but use rustfmt to get the file formatted (I believe).

Reply:
Macros and codegen are not the way to go IMO. This generator will have to generate C++ code in the future, so it really just needs to build a string for the target language, it's just a coincidence that it's written in Rust and also generates Rust.

This is more or less what flatc (and almost any other code generator) does already: 

https://github.com/google/flatbuffers/blob/9aa08a429e340bb8adf9b2e978ce2817cc8007c5/src/idl_gen_rust.cpp#L2038-L2042

The generated code is not covered by CI, like generated protobuf code is not covered by CI. Linting and formatting rules don't apply to generated code, since it's almost never checked in (we only check it in here to have a regression test, and it would be pointless to update the generator to emit code that is correctly formatted for our own CI). Adding a dependency on rustfmt for this generator seems overkill, especially since it will be used for other languages too. But even if it were possible, it would be pointless, since it's only generated in order to be used by the compiler (and the IDE perhaps), not to be read.

Reply:
Fair. I personally wouldn't mind using a macro, but I get your point about the generator & generated code only coincidentally sharing the same language. 

I do personally think that code should be generated with the prospect of people reading it in mind (I tend do that quite often when debugging, or even reviewing the inner working of a code generator). Hence it's imo a valid point. But don't feel super strongly about it, over the years I've learned how to read unreadable code :p.

Reply:
Yeah I see your point, maybe a macro that just helps with the indentation would be useful across languages. I'll look into that separately and see if it makes things more readable (in the generator).

But nicely formatting the generated code is not useful IMO (also because in general the user of the generated code is someone else who may have different formatting preferences). E.g. having the generator try to break lines at 80 characters, or running rustfmt afterwards are unnecessary complications and make the generator much less portable.

P.S.: Thanks for reading the unreadable generated code :upside_down_face:

Reply:
Also to clarify my previous point: I meant that it should not be covered by format / lint checks in CI. Correctness is actually checked in CI by means of the oak_idl_tests crate.